### ISSUE TITLE
Analyzer does not report "Method 'X' cannot be called on 'O' because it is potentially null"

### ISSUE THREAD

**Issue by kinex** [dart-lang/sdk#51156]
Code sample:

```dart
class A {
  A();
  void f() {}
}

A? a;

void main() {
  final f = a != null
      ? Future.value(a)
      : Future<A?>.delayed(const Duration(seconds: 1)).then((_) => A());
  f.then((value) => value.f());
}
```

IDE (VS Code) does not report errors, but trying to run the app gives the error:
```
Error: Method 'f' cannot be called on 'A?' because it is potentially null.
package:testapp37/main.dart:14
- 'A' is from 'package:testapp37/main.dart' ('lib/main.dart').
package:testapp37/main.dart:1
Try calling using ?. instead.
  f.then((value) => value.f());
```

Expected behavior is that I should see the error in IDE (VS Code Problems-tab) while editing the code.

This modified code works in IDE as expected:
```dart
class A {
  A();
  void f() {}
}

A? a;

void main() {
  final f = a != null
      ? Future<A?>.value(a)
      : Future<A?>.delayed(const Duration(seconds: 1)).then((_) => A());
  f.then((value) => value.f());
}
```

Dart SDK version: 2.19.0 (stable) on "windows_x64"

**Comment by lrhn** [comment]
Same lack of error in dartpad.dev, seems like a missing error from the analyzer.

**Comment by srawlins** [comment]
Still no error from the analyzer. :(

**Comment by lrhn** [comment]
It's a type inference difference, caused by the constructor `Future<T>.value([FutureOr<T>? value])` inferring different `T` values from a nullable argument type in analyzer and CFE.

I can't say which is correct, maybe @stereotype441 can. I'm pretty sure I've seen the issue before, though.

A more direct example:
```dart
import "dart:async";
class C<T> {
   C([FutureOr<T>? value]); // Inference disagrees on this one.
   C.notOpt(FutureOr<T>? value); // Inference disagrees on this one.
   C.plain([T? value]); // No problem with this one.
}
// Not just constructors.
C<T> funcOpt<T>([FutureOr<T>? value]) => C<T>(value);
C<T> func<T>(FutureOr<T>? value) => C<T>(value);

void main() {
  int? a;
  var c1 = C(a); // C<int?> in CFE, C<int> in analyzer.
  // c1.st<E<C<int>>>(); // Fails in CFE, OK in analyzer.
  c1.st<E<C<int?>>>(); // Fails in analyzer

  var c2 = C(a); // C<int?> in CFE, C<int> in analyzer.
  // c2.st<E<C<int>>>(); // Fails in CFE, OK in analyzer.
  c2.st<E<C<int?>>>(); // Fails in analyzer

  var c3 = C.plain(a); // C<int> everywhere
  c3.st<E<C<int>>>();

  var c4 = funcOpt(a); // C<int?> in CFE, C<int> in analyzer.
  // c4.st<E<C<int>>>(); // Fails in CFE, OK in analyzer.
  c4.st<E<C<int?>>>(); // Fails in analyzer

  var c5 = func(a); // C<int?> in CFE, C<int> in analyzer.
  // c5.st<E<C<int>>>(); // Fails in CFE, OK in analyzer.
  c5.st<E<C<int?>>>(); // Fails in analyzer
}

// Static type helpers.
extension <T> on T {
  void st<X extends E<T>>() { print("$this:$runtimeType @ $T"); }
}
typedef E<T> = T Function(T);
```

So it's a question of how a constraint of `int? <: FutureOr<T>?` solves for `T`.
Doesn't matter if it's optional, does matter that `FutureOr` is involved, without it `int? <: T?` does solve `T = int`.

I think the _analyzer might be doing the right thing_ here, or at least the consistent thing.

If `int? <: T?` makes `T` be `int`, why does `int? <: (T? | Future<T>)` not see that `int?` is never a `Future<T>`, and solve the same way for the remeaining `int? <: T?`?

(Is it the usual mistake of not checking for nullability before whether it's a `FutureOr` type, and using the `FutureOr<T>` type inference for a `FutureOr<T>?` type?)

If we have to decide who's right, I don't want to decide by what works best for `Future.value`, because it's a badly designed API for Dart 2. Here `int?` actually does work better. That suggests that a change could be breaking, but if the analyzer already rejects the program, we should be OK with changing behavior.

**Comment by lrhn** [comment]
> Is it the usual mistake of not checking for nullability ... ?

I think it is. Attempted fix: https://dart-review.googlesource.com/c/sdk/+/370642

There may still be a difference when the `FutureOr<T>?` occurs on the smaller side of the constraints:
```dart
import "dart:async";

void main() {
  FutureOr<T>? bar<T extends num>([Object? value]) => value as dynamic;
  FutureOr<Object>? o = Future<int>.value(0) as dynamic;
  if (o is FutureOr<int>) {
    // Both FutureOr<int> and FutureOr<int>? are types of interest.
    o = bar(o)
       ..st<E<FutureOr<int>?>>() // CFE OK
      // ..st<E<FutureOr<Object>?>>() // Analyzer OK
      ;
  }
}

// Static type helpers.
extension<T> on T {
  void st<X extends E<T>>() {
    print("$this:$runtimeType @ $T");
  }
}

typedef E<T> = T Function(T);
```
This gives different types after assignment to `o` in Analyzer and CFE:
* Analyzer: `FutureOr<num>?` (which is `T` instantiated to bounds).
* CFE: `FutureOr<int>?`.

That suggests that the CFE ignores the `?` and solves `FutureOr<T> <: FutureOr<int>`, then substitutes `int` back in and sees that `FutureOr<int>?` is not assignable.
The Analyzer probably looks at `FutureOr<T>? <: FutureOr<int>` and gives up, because a nullable type cannot be a subtype of a non-nullable type. With no solution to bind to `T`, it instantiates to bounds.

So, most likely one more missing nullability test. (Took a little to find a case which wasn't an error either way.)

**Comment by stereotype441** [comment]
Ok, let me rephrase what you're saying to make sure I understand it.

Future.value is defined like this:
```dart
abstract interface class Future<T> {
  factory Future.value([FutureOr<T>? value]) { ... }
  ...
}
```

Which means that if `x` has static type `A?`, then both `Future.value<A>(x)` and `Future.value<A?>(x)` are accepted by the type checker. But if the user writes `Future.value(x)`, then what they want is almost certainly `Future.value<A?>(x)` (because `Future.value<A>(x)` would throw an exception if `x` is `null`).

Today, things mostly work out ok, because of a previously unknown (at least to me) difference between CFE and analyzer type inference, which causes the CFE to treat `Future.value(x)` as `Future<A?>.value(x)`, while the analyzer treats it as `Future<A>.value(x)`. The crux of the issue is this line from https://github.com/dart-lang/language/blob/main/resources/type-system/inference.md#subtype-constraint-generation:

> If `Q` is `FutureOr<Q0>` the match holds under constraint set `C`:

The CFE is failing to check the nullability of `Q`, so it's effectively treating this as:

> If `Q` is `FutureOr<Q0>` or `FutureOr<Q0>?`, the match holds under constraint set `C`:

This interpretation is clearly not what the spec text intends, so it's tempting to change the CFE to match the spec, but that means that any use of `Future.value(...)` applied to a nullable argument is going to start throwing exceptions if its argument is `null`. Which would be a very bad breakage indeed.

Do I have that about right?

I think it's crucial that we not break existing code that uses `Future.value`. I can see a few ideas for how to proceed:

1. Change the analyzer (and the spec) to match what was implemented in the CFE. This is very tempting because it's easy and non-breaking. But I don't like it, because it feels like the story here is: "we came up with a bad API for `Future.value` before we had null safety, and then when we added null safety, instead of fixing it, we added a hack to the type system to avoid that bad API leading to bad type inference and subtle bugs. Sorry, users!"

2. Leave the CFE's type inference behavior as is for now, but deprecate the ability to pass zero arguments to `Future.value`. Immediately put out a lint and quick fix to encourage users to replace `Future.value()` with `Future.value(null)` and `Future<T>.value()` with `Future<T>.value(null)`. Then, in a future release, once users have had a chance to add explicit `null`s to their `Future.value()` calls, change the definition of `Future.value` to `factory Future.value(FutureOr<T> value) { ... }`. I like this approach because it fixes the bad API for `Future.value`, but I don't like the fact that it postpones the fix to CFE type inference until later; once we _do_ get around to making that fix, we might find out that there's some _other_ serious reason we can't make the fix, and then we'll be back to square one.

3. Fix the CFE's type inference behavior now, but at the same time, add a special case type inference rule to both the CFE and analyzer that causes `Future.value` with one argument to type infer as though it's declared `factory Future.value(FutureOr<T> value) { ... }`, and `Future.value` with zero arguments to type infer as though there's an invisible `null` argument. Then, in a future release, once users have had a chance to add explicit `null`s to their `Future.value()` calls, change the definition of `Future.value` to `factory Future.value(FutureOr<T> value) { ... }`, and drop the special case type inference rule. I like this approach because it fixes the bad API for `Future.value`, and it fixes the bad type inference behavior, and it gives us immediate feedback on whether this approach is going to work out in the long run. But it's also the most work of all three approaches.

I'm happy to consider other ideas too.

**Comment by lrhn** [comment]
> Do I have that about right?

Yep, precisely. 

I'd prefer option 2, because it fixed the bad API sooner, but we've lived with it so far, so we can probably take the time to do this correctly. 

I'd be perfectly happy to do a language versioned hack that makes `Future.value`  behave like a parameter type of `FutureOr<T>` (no `?`, not optional) for code that has language version 3.next and forward.
There will be migration, so a release first where it's only a warning may give a smoother transition.

We could perhaps make the fix language versioned too, to introduce it sooner.

(And we should check for more issues of this kind. As I wrote, there is at least one more.)
